internal/versioner/external.go

```diff
@@ -0,0 +1,104 @@
+// Copyright (C) 2014 The Syncthing Authors.
+//
+// This program is free software: you can redistribute it and/or modify it
+// under the terms of the GNU General Public License as published by the Free
+// Software Foundation, either version 3 of the License, or (at your option)
+// any later version.
+//
+// This program is distributed in the hope that it will be useful, but WITHOUT
+// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+// FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
+// more details.
+//
+// You should have received a copy of the GNU General Public License along
+// with this program. If not, see <http://www.gnu.org/licenses/>.
+
+package versioner
+
+import (
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+)
+
+func init() {
+	// Register the constructor for this type of versioner with the name "external"
+	Factories["external"] = NewExternal
+}
+
+// The type holds our configuration
+type External struct {
+	command    string
+	folderPath string
+}
+
+// The constructor function takes a map of parameters and creates the type.
+func NewExternal(folderID, folderPath string, params map[string]string) Versioner {
+	command := params["command"]
+
+	s := External{
+		command:    command,
+		folderPath: folderPath,
+	}
+
+	if debug {
+		l.Debugf("instantiated %#v", s)
+	}
+	return s
+}
+
+// Move away the named file to a version archive. If this function returns
+// nil, the named file does not exist any more (has been archived).
+func (v External) Archive(filePath string) error {
+	_, err := os.Lstat(filePath)
+	if err != nil {
+		if os.IsNotExist(err) {
+			if debug {
+				l.Debugln("not archiving nonexistent file", filePath)
+			}
+			return nil
+		}
+		return err
+	}
+
+	if debug {
+		l.Debugln("archiving", filePath)
+	}
+
+	inFolderPath, err := filepath.Rel(v.folderPath, filePath)
+	if err != nil {
+		return err
+	}
+	if v.command != "" {
+		cmd := exec.Command(v.command, v.folderPath, inFolderPath)
+		env := os.Environ()
+		// filter STGUIAUTH and STGUIAPIKEY from environment variables
+		fiteredEnv := []string{}
+		for _, x := range env {
+			if !strings.HasPrefix(x, "STGUIAUTH=") && !strings.HasPrefix(x, "STGUIAPIKEY=") {
+				fiteredEnv = append(fiteredEnv, x)
+			}
+		}
+		cmd.Env = fiteredEnv
+		err = cmd.Run()
+		if err != nil {
+			l.Warnln("Versioner:", err)
+		}
+	} else {
+		l.Warnln("Versioner: command is empty, please enter a valid command")
```

Comment:
so after every call of `Archive` we do something like
```
if err != nil {
	l.Infoln("versioner error:", err)
	err=os.Remove(file)
}
```
in the puller to be sure that syncthing can continue? Like this we can continue, but versions could be lost...
I don't know what is better here, making progress and losing versions (and the user will maybe not even notice if it's just a info) or stopping and spamming warnings, with no versions lost. I think I actually prefer stopping because there is something wrong that the user should fix if we get errors.